<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Internals · PolyOrigin.jl</title><link rel="canonical" href="https://chaozhi.github.io/PolyOrigin.jl/lib/internals/"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit">PolyOrigin.jl</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../../man/0_get_started/">0 Get started</a></li><li><a class="tocitem" href="../../man/1_prepare_input/">1 Prepare input</a></li><li><a class="tocitem" href="../../man/2_run_polyOrigin/">2 Run polyOrigin</a></li><li><a class="tocitem" href="../../man/3_examples/">3 Examples</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../public/">Public</a></li><li class="is-active"><a class="tocitem" href>Internals</a><ul class="internal"><li><a class="tocitem" href="#Contents"><span>Contents</span></a></li><li><a class="tocitem" href="#Index"><span>Index</span></a></li><li><a class="tocitem" href="#Internal-Interface"><span>Internal Interface</span></a></li></ul></li></ul></li><li><a class="tocitem" href="../../symbols/">Index</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Library</a></li><li class="is-active"><a href>Internals</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Internals</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/chaozhi/PolyOrigin.jl/blob/master/docs/src/lib/internals.md#L" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="PolyOrigin-Documentation"><a class="docs-heading-anchor" href="#PolyOrigin-Documentation">PolyOrigin Documentation</a><a id="PolyOrigin-Documentation-1"></a><a class="docs-heading-anchor-permalink" href="#PolyOrigin-Documentation" title="Permalink"></a></h1><p>Documentation for <code>PolyOrigin.jl</code>&#39;s internal interface.</p><h2 id="Contents"><a class="docs-heading-anchor" href="#Contents">Contents</a><a id="Contents-1"></a><a class="docs-heading-anchor-permalink" href="#Contents" title="Permalink"></a></h2><ul><li><a href="#PolyOrigin-Documentation">PolyOrigin Documentation</a></li><ul><li><a href="#Contents">Contents</a></li><li><a href="#Index">Index</a></li><li><a href="#Internal-Interface">Internal Interface</a></li></ul></ul><h2 id="Index"><a class="docs-heading-anchor" href="#Index">Index</a><a id="Index-1"></a><a class="docs-heading-anchor-permalink" href="#Index" title="Permalink"></a></h2><ul><li><a href="#PolyOrigin.polyMapRefine!"><code>PolyOrigin.polyMapRefine!</code></a></li><li><a href="#PolyOrigin.polyPhase"><code>PolyOrigin.polyPhase</code></a></li><li><a href="#PolyOrigin.polyPhase!"><code>PolyOrigin.polyPhase!</code></a></li><li><a href="#PolyOrigin.polyReconstruct!"><code>PolyOrigin.polyReconstruct!</code></a></li><li><a href="#PolyOrigin.savegenodata"><code>PolyOrigin.savegenodata</code></a></li><li><a href="#PolyOrigin.savegenoprob"><code>PolyOrigin.savegenoprob</code></a></li><li><a href="#PolyOrigin.setAbsPhase!"><code>PolyOrigin.setAbsPhase!</code></a></li></ul><h2 id="Internal-Interface"><a class="docs-heading-anchor" href="#Internal-Interface">Internal Interface</a><a id="Internal-Interface-1"></a><a class="docs-heading-anchor-permalink" href="#Internal-Interface" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="PolyOrigin.polyPhase" href="#PolyOrigin.polyPhase"><code>PolyOrigin.polyPhase</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">polyPhase(genofile,pedfile, delimchar=&#39;,&#39;, missingstring=&quot;NA&quot;,
    commentstring=&quot;#&quot;, keyargs...)</code></pre><p>performs parental phasing and return phasedgeno::PolyGeno with phasedgeno.parentgeno being phased.</p><p><strong>Positional arguments</strong></p><p><code>genofile::AbstractString</code>: filename for genotypic data file.</p><p><code>pedfile::AbstractString</code>:  filename for population pedigree.</p><p>see <a href="../public/#PolyOrigin.readPolyGeno"><code>readPolyGeno</code></a> for the requirements of <code>genofile</code> and <code>pedfile</code>.</p><p><strong>Keyword arguments</strong></p><p><code>delimchar::AbstractChar=&#39;,&#39;</code>:  text delimiter.</p><p><code>missingstring::AbstractString=&quot;NA&quot;</code>: string code for missing value.</p><p><code>commentstring::AbstractString=&quot;#&quot;</code>: rows that begins with commentstring will be ignored.</p><p>see keyargs in polyPhase!(polygeno::PolyGeno, keyargs...)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/chaozhi/PolyOrigin.jl/blob/c5e4e897a0e05197b33611817639d3c515a3ffcf/src/polyphase.jl#LL46-L71">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PolyOrigin.polyPhase!" href="#PolyOrigin.polyPhase!"><code>PolyOrigin.polyPhase!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">polyPhase!(polygeno::PolyGeno, keyargs...)</code></pre><p>performs parental phasing and return phasedgeno::PolyGeno with  phasedgeno.parentgeno and polygeno.parentgeno being phased .</p><p><strong>Positional arguments</strong></p><p><code>polygeno::PolyGeno</code>: a struct that stores genotypic data and pedigree info.</p><p><strong>Keyword arguments</strong></p><p><code>epsilon::Real=0.01</code>: genotypic error probability.</p><p><code>seqerror::Real=0.001</code>: base sequencing error probability for GBS data.</p><p><code>chrpairing_phase::Integer=22</code>: chromosome pairing in parental phasing, with 22 being only bivalent formations and 44 being bi- and quadri-valent formations.</p><p><code>chrsubset::Union{Nothing,AbstractRange,AbstractVector}=nothing</code>: subset of chromosomes to be considered, with nothing denoting all chromosomes. Delete chromosome indices that are out of range.</p><p><code>snpsubset::Union{Nothing,AbstractRange,AbstractVector}=nothing</code>: subset of markers to be considered, with nothing denoting all markers. within a chromosome, marker index starts from 1, and marker indices that are larger than the number of markers within the chromosome are deleted.</p><p><code>isparallel::Bool=false</code>: if true, multicore computing over chromosomes.</p><p><code>delmarker::Bool=true</code>: if true, delete markers during parental phasing.</p><p>&quot;delsiglevel::Real=0.05&quot;: significance level for deleting markers.</p><p><code>maxstuck::Integer=5</code>: the max number of consecutive iterations that are rejected in a phasing run.</p><p><code>maxiter::Integer=30</code>: the max number of iterations in a phasing run.&quot;</p><p><code>minrun::Integer=3</code>: if the min number of phasing runs that are at the same local maximimum or have the same parental phases reaches minrun, phasing algorithm will stop before reaching the maxrun.</p><p><code>maxrun::Integer=10</code>: the max number of phasing runs.</p><p><code>byparent::Union{Nothing,Bool}=nothing</code>: if true, update parental phases  parent by parent; if false, update parental phases one subpopulation by subpopulation.  The nothing denotes that it is true if a connected component is a simple F1 cross, and false otherwise.</p><p><code>byneighbor::Union{Nothing,Bool}=nothing</code>: if ture, udpate the combination of bivalent or multivalents in parents by their neighbors; if false, consider all the possible combinations. The nothing denotes thtat it is true if max ploidy&gt;=6, and false otherwise.</p><p><code>refhapfile::Union{Nothing,AbstractString} = nothing</code>: reference haplotype file for setting absolute parental phases. It has the same format as the input genofile, except that parental genotypes are phased and offspring genotypes are ignored if they exist.</p><p><code>missingstring::AbstractString=&quot;NA&quot;</code>: string code for missing value.</p><p><code>outstem::Union{Nothing,AbstractString}=&quot;outstem&quot;</code>: stem of output filenames. If nothing, no output files.</p><p><code>logfile::Union{Nothing,AbstractString,IO}= (outstem===nothing ? nothing : string(outstem,&quot;.log&quot;))</code>: log file or IO for writing log. If nothing, no log file.</p><p><code>workdir::AbstractString = pwd()</code>: directory for reading and writing files.</p><p><code>verbose::Bool=true</code>: if true, print messages on console.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/chaozhi/PolyOrigin.jl/blob/c5e4e897a0e05197b33611817639d3c515a3ffcf/src/polyphase.jl#LL111-L180">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PolyOrigin.polyMapRefine!" href="#PolyOrigin.polyMapRefine!"><code>PolyOrigin.polyMapRefine!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">polyMapRefine!(phasedgeno::PolyGeno, keyargs...)</code></pre><p>performs marker map refinning for phasedgeno with phased parent genotypes.     Modifies phasedgeno.markermap into a refined genetic map.</p><p><strong>Positional arguments</strong></p><p><code>phasedgeno::PolyGeno</code>: a struct that stores genotypic data and pedigree info.</p><p><strong>Keyword arguments</strong></p><p><code>epsilon::Real=0.01</code>: genotypic error probability for offspring phasing.</p><p><code>chrpairing::Integer=44</code>: chromosome pairing in offspring decoding, with 22 being only bivalent formations and 44 being bivalent and quadrivalent formations.</p><p><code>chrsubset::Union{Nothing,AbstractRange,AbstractVector}=nothing</code>: subset of chromosomes to be considered, with nothing denoting all chromosomes. Delete chromosome indices that are out of range.</p><p><code>snpsubset::Union{Nothing,AbstractRange,AbstractVector}=nothing</code>: subset of markers to be considered, with nothing denoting all markers. within a chromosome, marker index starts from 1, and marker indices that are larger than the number of markers within the chromosome are deleted.</p><p><code>isparallel::Bool=false</code>: if true, multicore computing over chromosomes.</p><p><code>refineorder::Bool=false</code>: if true, refine marker mordering.</p><p><code>maxwinsize::Integer=50</code>: max size of sliding windown in map refinning.</p><p><code>inittemperature::Real=4</code>: initial temperature of simulated annealing in map refinning.</p><p><code>coolingrate::Real=0.5</code>: cooling rate of annealing temperature in map refinning.</p><p><code>stripdis::Real=20</code>: a chromosome end in map refinement is removed if it has a distance gap &gt; stripdis (centiMorgan) and it contains less than 5% markers.</p><p><code>maxepsilon::Real=0.5</code>: markers in map refinement are removed it they have error rates &gt; maxepsilon.</p><p><code>skeletonsize::Integer=50</code>: the number of markers in the skeleton map that is used to re-scale inter-map distances.</p><p><code>logfile::Union{AbstractString,IOStream}=string(outstem,&quot;.log&quot;)</code>: output filenames or stream for writing log.</p><p><code>workdir::AbstractString = pwd()</code>: directory for reading and writing files.</p><p><code>verbose::Bool=true</code>: if true, print messages on console.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/chaozhi/PolyOrigin.jl/blob/c5e4e897a0e05197b33611817639d3c515a3ffcf/src/polymaprefine.jl#LL862-L914">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PolyOrigin.polyReconstruct!" href="#PolyOrigin.polyReconstruct!"><code>PolyOrigin.polyReconstruct!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">polyReconstruct!(polygeno::PolyGeno, keyargs...)</code></pre><p>performs ancestral inference from polygeno and return polyancestry::PolyAncestry.</p><p><strong>Positional arguments</strong></p><p><code>polygeno::PolyGeno</code>: a struct that stores genotypic data and pedigree info.</p><p><strong>Keyword arguments</strong></p><p><code>epsilon::Real=0.01</code>: genotypic error probability for offspring phasing.</p><p><code>seqerror::Real=0.001</code>: base sequencing error probability for GBS data.</p><p><code>chrpairing::Integer=44</code>: chromosome pairing in offspring decoding, with 22 being only bivalent formations and 44 being bivalent and quadrivalent formations.</p><p><code>chrsubset::Union{Nothing,AbstractRange,AbstractVector}=nothing</code>: subset of chromosomes to be considered, with nothing denoting all chromosomes. Delete chromosome indices that are out of range.</p><p><code>snpsubset::Union{Nothing,AbstractRange,AbstractVector}=nothing</code>: subset of markers to be considered, with nothing denoting all markers. within a chromosome, marker index starts from 1, and marker indices that are larger than the number of markers within the chromosome are deleted.</p><p><code>isparallel::Bool=false</code>: if true, multicore computing over chromosomes.</p><p><code>correctthreshold::AbstractFloat=0.15</code>: a candidate marker is selected for parental error correction if the fraction of offspring genotypic error &gt;= correctthreshold.</p><p><code>isplot::Bool=false</code>: if true, plot haploprob for all offspring and save in the folder &quot;outstem_plots&quot;.</p><p><code>outstem::Union{Nothing,AbstractString}=&quot;outstem&quot;</code>: stem of output filenames. If nothing, no output files.</p><p><code>logfile::Union{Nothing,AbstractString,IO}= (outstem===nothing ? nothing : string(outstem,&quot;.log&quot;))</code>: log file or IO for writing log. If nothing, no log file.</p><p><code>workdir::AbstractString = pwd()</code>: directory for reading and writing files.</p><p><code>verbose::Bool=true</code>: if true, print messages on console.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/chaozhi/PolyOrigin.jl/blob/c5e4e897a0e05197b33611817639d3c515a3ffcf/src/polyreconstruct.jl#LL1-L46">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PolyOrigin.setAbsPhase!" href="#PolyOrigin.setAbsPhase!"><code>PolyOrigin.setAbsPhase!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">setAbsPhase!(refhapfile,phasedgeno::PolyGeno,io=nothing,verbose=true)</code></pre><p>set absolute parental phases, based on the reference haplotype file.</p><p><strong>Positional arguments</strong></p><p><code>refhapfile::AbstractString</code>: reference haplotype file has the same format as the input genofile, except that parental genotypes are phased and offspring genotypes are ignored if they exist.</p><p><code>phasedgeno::PolyGeno</code>: polygeno struct with phased parent genotypes .</p><p><strong>Keyward arguments</strong></p><p><code>workdir::AbstractString = pwd()</code>: directory for reading <code>refhapfile</code>.</p><p><code>io::Union{Nothing,IOStream}</code>: stream for writing log.</p><p><code>verbose::Bool=true</code>: true if print messages on console.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/chaozhi/PolyOrigin.jl/blob/c5e4e897a0e05197b33611817639d3c515a3ffcf/src/setabsphase.jl#LL246-L267">source</a></section><section><div><pre><code class="language-none">setAbsPhase!(refhapfile, polyancestry,io=nothing,verbose=true)</code></pre><p>set absolute parental phases and consistent polyancestry.genoprob and polyancestry.haploprob, based on the reference haplotype file.</p><p><strong>Positional arguments</strong></p><p><code>refhapfile::AbstractString</code>: reference haplotype file has the same format as the input genofile, except that parental genotypes are phased and offspring genotypes are ignored if they exist.</p><p><code>polyancestry::PolyAncestry</code>: polyancestry struct with phased parent genotypes and inference genoprob.</p><p><strong>Keyward arguments</strong></p><p><code>workdir::AbstractString = pwd()</code>: directory for reading <code>trueancestryfile</code>.</p><p><code>io::Union{Nothing,IOStream}</code>: stream for writing log.</p><p><code>verbose::Bool=true</code>: true if print messages on console.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/chaozhi/PolyOrigin.jl/blob/c5e4e897a0e05197b33611817639d3c515a3ffcf/src/setabsphase.jl#LL288-L310">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PolyOrigin.savegenodata" href="#PolyOrigin.savegenodata"><code>PolyOrigin.savegenodata</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">savegenodata(outfile,polygeno,missingstring=&quot;NA&quot;,workdir=pwd())</code></pre><p>saves genotypic data of the struct polygeno into outfile</p><p><strong>Positional arguments</strong></p><p><code>outfile::AbstractString</code>: filename for saving results.</p><p><code>polygeno::PolyGeno</code>: a struct returned by <a href="../public/#PolyOrigin.readPolyGeno"><code>readPolyGeno</code></a>.</p><p><strong>Keyward arguments</strong></p><p><code>missingstring::AbstractString=&quot;NA&quot;</code>: string code for missing value.</p><p><code>workdir::AbstractString = pwd()</code>: directory for writing outfile.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/chaozhi/PolyOrigin.jl/blob/c5e4e897a0e05197b33611817639d3c515a3ffcf/src/polygeno.jl#LL443-L460">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PolyOrigin.savegenoprob" href="#PolyOrigin.savegenoprob"><code>PolyOrigin.savegenoprob</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">savegenoprob(outfile,polyancestry,missingstring=&quot;NA&quot;,workdir=pwd())</code></pre><p>saves polyancestry.genoprob and polyancestry.parentgeno into outfile.</p><p><strong>Positional arguments</strong></p><p><code>outfile::AbstractString</code>: filename for saving results.</p><p><code>polyancestry::PolyAncestry</code>: a struct returned by <a href="../public/#PolyOrigin.polyOrigin"><code>polyOrigin</code></a>.</p><p><strong>Keyward arguments</strong></p><p><code>missingstring::AbstractString=&quot;NA&quot;</code>: string code for missing value.</p><p><code>workdir::AbstractString = pwd()</code>: directory for writing outfile.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/chaozhi/PolyOrigin.jl/blob/c5e4e897a0e05197b33611817639d3c515a3ffcf/src/polyancestry.jl#LL562-L579">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../public/">« Public</a><a class="docs-footer-nextpage" href="../../symbols/">Index »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Tuesday 23 February 2021 17:44">Tuesday 23 February 2021</span>. Using Julia version 1.5.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
