<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Public · PolyOrigin.jl</title><link rel="canonical" href="https://chaozhi.github.io/PolyOrigin.jl/lib/public/"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit">PolyOrigin.jl</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../../man/0_get_started/">0 Get started</a></li><li><a class="tocitem" href="../../man/1_prepare_input/">1 Prepare input</a></li><li><a class="tocitem" href="../../man/2_run_polyOrigin/">2 Run polyOrigin</a></li><li><a class="tocitem" href="../../man/3_examples/">3 Examples</a></li></ul></li><li><span class="tocitem">Library</span><ul><li class="is-active"><a class="tocitem" href>Public</a><ul class="internal"><li><a class="tocitem" href="#Contents"><span>Contents</span></a></li><li><a class="tocitem" href="#Index"><span>Index</span></a></li><li><a class="tocitem" href="#Public-Interface"><span>Public Interface</span></a></li></ul></li><li><a class="tocitem" href="../internals/">Internals</a></li></ul></li><li><a class="tocitem" href="../../symbols/">Index</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Library</a></li><li class="is-active"><a href>Public</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Public</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/chaozhi/PolyOrigin.jl/blob/master/docs/src/lib/public.md#L" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="PolyOrigin-Documentation"><a class="docs-heading-anchor" href="#PolyOrigin-Documentation">PolyOrigin Documentation</a><a id="PolyOrigin-Documentation-1"></a><a class="docs-heading-anchor-permalink" href="#PolyOrigin-Documentation" title="Permalink"></a></h1><p>Documentation for <code>PolyOrigin.jl</code>&#39;s public interface.</p><p>See the internals section for internal function docs.</p><h2 id="Contents"><a class="docs-heading-anchor" href="#Contents">Contents</a><a id="Contents-1"></a><a class="docs-heading-anchor-permalink" href="#Contents" title="Permalink"></a></h2><ul><li><a href="#PolyOrigin-Documentation">PolyOrigin Documentation</a></li><ul><li><a href="#Contents">Contents</a></li><li><a href="#Index">Index</a></li><li><a href="#Public-Interface">Public Interface</a></li></ul></ul><h2 id="Index"><a class="docs-heading-anchor" href="#Index">Index</a><a id="Index-1"></a><a class="docs-heading-anchor-permalink" href="#Index" title="Permalink"></a></h2><ul><li><a href="#PolyOrigin"><code>PolyOrigin</code></a></li><li><a href="#PolyOrigin.PolyAncestry"><code>PolyOrigin.PolyAncestry</code></a></li><li><a href="#PolyOrigin.PolyGeno"><code>PolyOrigin.PolyGeno</code></a></li><li><a href="#PolyOrigin.animCondprob"><code>PolyOrigin.animCondprob</code></a></li><li><a href="#PolyOrigin.calAccuracy!"><code>PolyOrigin.calAccuracy!</code></a></li><li><a href="#PolyOrigin.plotCondprob"><code>PolyOrigin.plotCondprob</code></a></li><li><a href="#PolyOrigin.plotMapComp"><code>PolyOrigin.plotMapComp</code></a></li><li><a href="#PolyOrigin.polyOrigin"><code>PolyOrigin.polyOrigin</code></a></li><li><a href="#PolyOrigin.polyOrigin!"><code>PolyOrigin.polyOrigin!</code></a></li><li><a href="#PolyOrigin.readPolyAncestry"><code>PolyOrigin.readPolyAncestry</code></a></li><li><a href="#PolyOrigin.readPolyGeno"><code>PolyOrigin.readPolyGeno</code></a></li><li><a href="#PolyOrigin.readTruegeno!"><code>PolyOrigin.readTruegeno!</code></a></li><li><a href="#PolyOrigin.savePolyAncestry"><code>PolyOrigin.savePolyAncestry</code></a></li></ul><h2 id="Public-Interface"><a class="docs-heading-anchor" href="#Public-Interface">Public Interface</a><a id="Public-Interface-1"></a><a class="docs-heading-anchor-permalink" href="#Public-Interface" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="PolyOrigin" href="#PolyOrigin"><code>PolyOrigin</code></a> — <span class="docstring-category">Module</span></header><section><div><p>PolyOrigin is a package for haplotype reconstruction in connected polyploid F1 populations. See also: <a href="#PolyOrigin.polyOrigin"><code>polyOrigin</code></a>, <a href="#PolyOrigin.polyOrigin!"><code>polyOrigin!</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/chaozhi/PolyOrigin.jl/blob/906b6a5286692924a9e2b7f2c05060d964e3bebc/src/PolyOrigin.jl#LL3-L6">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PolyOrigin.PolyGeno" href="#PolyOrigin.PolyGeno"><code>PolyOrigin.PolyGeno</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">PolyGeno</code></pre><p>mutable struct that stores genotypic data and pedigree information</p><pre><code class="language-none">PolyGeno(markermap,parentgeno,offspringgeno,parentinfo,offspringinfo,designinfo,delmarker,correction)</code></pre><p>constructor from genotypic data and pedigree information.</p><p><strong>Fields</strong></p><p><code>markermap::Vector{DataFrame}</code>: marker map for each chromosome. <code>markermap[c]</code> gives the dataframe of chromosome c with columns [:marker, :chromosome, :position].</p><p><code>parentgeno::Vector{Matrix}</code>: genotypic data in parents. <code>parentgeno[c][m,i]</code> gives the genotype of parent i at marker m of chromosome c.</p><p><code>offspringgeno::Vector{Matrix}</code>: genotypic data in offspring. <code>offspringgeno[c][m,i]</code>  gives the genotype of offspring i at marker m of chomosome c.</p><p><code>parentinfo::DataFrame</code>: parent information with columns [:individual, :ploidy].</p><p><code>offspringinfo::DataFrame</code>: offspring information with columns [:individual,:population,:isoutlier].</p><p><code>designinfo::DataFrame</code>: design information with columns [:population, :parent1, :parent2, ...]. The dataframe element <code>n_ij</code> denotes the nubmer of gametes contributed to each offspring in i-th population by j-th parent. <code>n_ij</code>=2 means that i-th population is produced by self-fertilization of j-th parent. Row sum must be 2.</p><p><code>delmarker::DataFrame</code>: dataframe for collecting markers that were removed from markermap, in the stages of parental phasing or map refinement. It has columns [:marker, :chromosome, :position].</p><p><code>correction::DataFrame</code>: dataframe for collecting parental genotype corrections. It has columns [:round,:marker,:chromosome,:parent,:old<em>genotype,:new</em>genotype,:old<em>nerr,:new</em>nerr].</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>A genotype in <code>offspringgeno</code> must be one of following formats 1-3, and a genotype in <code>parentgeno</code> must be one of following formats 1-4:</p><ol><li><code>dosage</code>: ranges from 0, 1, ..., ploidy, and missing for missing dosage;</li><li><code>readcount</code>:  [c1, c2] where c1 and c2 are the number of reads for  alleles 1 and 2, respectively. Missing genotypes are denoted by [0,0].</li><li><code>probability</code>: [p0, p1, ...], where p_i denotes the probability of observed data given dosage i = 0, ..., ploidy, and the probabilities are normalized so that their sum is 1.</li><li><code>phasedgeno</code>: [g1, g2,...],  where g_i=1 or 2 for i=1,..., ploidy.</li></ol></div></div></div><a class="docs-sourcelink" target="_blank" href="https://github.com/chaozhi/PolyOrigin.jl/blob/906b6a5286692924a9e2b7f2c05060d964e3bebc/src/polygeno.jl#LL1-L51">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PolyOrigin.PolyAncestry" href="#PolyOrigin.PolyAncestry"><code>PolyOrigin.PolyAncestry</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">PolyAncestry</code></pre><p>mutable struct the results of haplotype reconstruction.</p><pre><code class="language-none">PolyAncestry(markermap,parentgeno,parentinfo,offspringinfo,designinfo,
    statespace,valentprob,genoprob)</code></pre><p>Constructor from the results of haplotype reconstruction.</p><p><strong>Fields</strong></p><p><code>markermap::Vector{DataFrame}</code>: marker map for each chromosome. <code>markermap[c]</code> gives the dataframe of chromosome c with columns [:marker, :chromosome, :position].</p><p><code>parentgeno::Vector{Matrix}</code>: genotypic data in parents. <code>parentgeno[c][m,i]</code> gives the genotype of parent i at marker m of chromosome c.</p><p><code>parentinfo::DataFrame</code>: parent information with columns [:individual, :ploidy].</p><p><code>offspringinfo::DataFrame</code>: offspring information with columns [:individual,:population].</p><p><code>designinfo::DataFrame</code>: design information with columns [:population, :parent1, :parent2, ...]. The dataframe element <code>n_ij</code> denotes the nubmer of gametes contributed to each offspring in i-th population by j-th parent. <code>n_ij</code>=2 means that i-th population is produced by self-fertilization of j-th parent. Row sum must be 2.</p><p><code>delmarker::DataFrame</code>: dataframe for collecting markers that were removed from markermap, in the stages of parental phasing or map refinement. It has columns [:marker, :chromosome, :position].</p><p><code>correction::DataFrame</code>: dataframe for collecting parental genotype corrections. It has columns [:marker,:chromosome,:parent,:old<em>genotype,:new</em>genotype,:old<em>nerr,:new</em>nerr].</p><p><code>statespace::Dict</code>: specify valents and origin-genotypes for each population. Each key is a population id, and its value is in turn a dict with keys: &quot;parent&quot;, &quot;parentindex&quot;, &quot;valent&quot;, &quot;valentneighbor&quot;, and &quot;groupstate&quot;.</p><p><code>valentprob::Union{Nothing,Vector}</code>, posterior valent probability for each offspring in each chromosome. valentprob[c][o] for offspring o in chromsome c is a matrix with three columns being valentindex, loglike, and posterior probability. Here posterior probability is calcuated from loglike, assuming a discrete uniform prior distribution of valent configurations.</p><p><code>genoprob::Vector</code>: posterior origin-genotype probability for each offspring in each chromosome. genoprob[c][o] for offspring o in chromsome c is a sparse matrix, with element (m,s) being the probability of origin-genotype s at marker m.</p><p><code>haploprob::Union{Nothing,Vector}</code>: posterior origin-haplotype probability for each offspring in each chromosome. haploprob[c][o] for offspring o in chromsome c is a sparse matrix, with element (m,s) being the probability of origin-haplotype s at marker m.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/chaozhi/PolyOrigin.jl/blob/906b6a5286692924a9e2b7f2c05060d964e3bebc/src/polyancestry.jl#LL1-L54">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PolyOrigin.polyOrigin" href="#PolyOrigin.polyOrigin"><code>PolyOrigin.polyOrigin</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">polyOrigin(genofile, pedfile, delimchar=&#39;,&#39;, missingstring=&quot;NA&quot;,
    commentstring=&quot;#&quot;, keyargs...)</code></pre><p>performs parental phasing and ancestral inference from input files and return polyancestry::PolyAncestry. Only ancestral inference is performed in the case of phased parents.</p><p><strong>Positional arguments</strong></p><p><code>genofile::AbstractString</code>: filename for genotypic data file.</p><p><code>pedfile::AbstractString</code>:  filename for pedigree info.</p><p>see <a href="#PolyOrigin.readPolyGeno"><code>readPolyGeno</code></a> for the requirements of <code>genofile</code> and <code>pedfile</code>.</p><p><strong>Keyword arguments</strong></p><p><code>delimchar::AbstractChar=&#39;,&#39;</code>:  text delimiter.</p><p><code>missingstring::AbstractString=&quot;NA&quot;</code>: string code for missing value.</p><p><code>commentstring::AbstractString=&quot;#&quot;</code>: rows that begin with commentstring will be ignored.</p><p><code>isphysmap::Bool=false</code>: if true, input markermap is physical map, where marker locations are in unit of base pair(bp).</p><p><code>recomrate::Real=1</code>: recombination rate in unit of 1 cM/Mbp (centiMorgan per million base pair). Valid only if <code>isphysmap=true</code>.</p><p>see keyargs in polyOrigin!(polygeno::PolyGeno, keyargs...)</p><p><strong>Output files</strong></p><p><code>outstem.log</code>: log file saves messages that were printed on console.</p><p><code>outstem_maprefined.csv</code>: same as the input genofile, except that input marker map is refined.</p><p><code>outstem_parentphased.csv</code>: same as the input genofile, except that parental genotypes are phased.</p><p><code>outstem_parentphased_corrected.csv</code>: parented genotypes are further corrected.</p><p><code>outstem_polyancestry.csv</code>: saves the returned polyancestry. See <a href="#PolyOrigin.savePolyAncestry"><code>savePolyAncestry</code></a>.</p><p><code>outstem_genoprob.csv</code>: a concise version of the above file, including genetic map, phased parental genotypes, and posterior origin-genotype probabilities. See <a href="../internals/#PolyOrigin.savegenoprob"><code>savegenoprob</code></a>.</p><p><code>outstem_postdoseprob.csv</code>: same as the input genofile, except that  parent genotypes are phased and offspring genotypes are given by the posterior dose probabilities.</p><p><strong>Example</strong></p><pre><code class="language-julia-repl">julia&gt; polyOrigin(genofile,pedfile)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/chaozhi/PolyOrigin.jl/blob/906b6a5286692924a9e2b7f2c05060d964e3bebc/src/polyorigin_fun.jl#LL2-L56">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PolyOrigin.polyOrigin!" href="#PolyOrigin.polyOrigin!"><code>PolyOrigin.polyOrigin!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">polyOrigin!(polygeno::PolyGeno, keyargs...)</code></pre><p>performs parental phasing and ancestral inference from polygeno and return polyancestry::PolyAncestry. Only ancestral inference is performed in the case of phased parents. See <a href="#PolyOrigin.readPolyGeno"><code>readPolyGeno</code></a> for creating polygeno from inputfiles.</p><p><strong>Positional arguments</strong></p><p><code>polygeno::PolyGeno</code>:  a struct storing genotypic data and pedigree info.</p><p><strong>Keyword arguments</strong></p><p><code>epsilon::Real=0.01</code>: genotyping error probability.</p><p><code>seqerr::Real=0.001</code>: sequencing read error probability for GBS data.</p><p><code>chrpairing_phase::Integer=22</code>: chromosome pairing in parental phasing, with 22 being only bivalent formations and 44 being bivalent and quadrivalent formations.</p><p><code>chrpairing::Integer=44</code>: chromosome pairing in offspring decoding, with 22 being only bivalent formations and 44 being bivalent and quadrivalent formations.</p><p><code>chrsubset::Union{Nothing,AbstractRange,AbstractVector}=nothing</code>: subset of chromosome, with nothing denoting all chromosomes. Delete chromosome indices that are out of range.</p><p><code>snpsubset::Union{Nothing,AbstractRange,AbstractVector}=nothing</code>: subset of markers to be considered, with nothing denoting all markers. within a chromosome, marker index starts from 1, and marker indices that are larger than the number of markers within the chromosome are deleted.</p><p><code>isparallel::Bool=false</code>: if true, multicore computing over chromosomes.</p><p><code>delmarker::Bool=true</code>: if true, delete markers during parental phasing.</p><p><code>delsiglevel::Real=0.05</code>: significance level for deleting markers.</p><p><code>maxstuck::Integer=5</code>: the max number of consecutive iterations that are rejected in a phasing run.</p><p><code>maxiter::Integer=30</code>: the max number of iterations in a phasing run.</p><p><code>minrun::Integer=3</code>: if the number of phasing runs having the same parental phases reaches minrun, phasing algorithm will stop before reaching the maxrun.</p><p><code>maxrun::Integer=10</code>: the max number of phasing runs.</p><p><code>byparent::Union{Nothing,Bool}=nothing</code>: if true, update parental phases parent by parent; if false, update parental phases one subpopulation by subpopulation. The nothing denotes that it is true if a connected component is a single F1 cross, and false otherwise.</p><p><code>byneighbor::Union{Nothing,Bool}=nothing</code>: if ture, udpate the combination of bivalent or multivalents in parents by their neighbors; if false, consider all the possible combinations. The nothing denotes that it is true if max ploidy&gt;=6, and false otherwise.</p><p><code>refhapfile::Union{Nothing,AbstractString} = nothing</code>: reference haplotype file for setting absolute parental phases. It has the same format as the input genofile, except that parental genotypes are phased and offspring genotypes are ignored if exist.</p><p><code>correctthreshold::AbstractFloat=0.15</code>: a candidate marker is selected for parental error correction if the fraction of offspring genotypic error &gt;= correctthreshold.</p><p><code>refinemap::Bool=false</code>: if true, refine marker map.</p><p><code>refineorder::Bool=false</code>: if true, refine marker mordering, valid only if refinemap=true</p><p><code>maxwinsize::Integer=50</code>: max size of sliding windown in map refinning.</p><p><code>inittemperature::Real=4</code>: initial temperature of simulated annealing in map refinning.</p><p><code>coolingrate::Real=0.5</code>: cooling rate of annealing temperature in map refinning.</p><p><code>stripdis::Real=20</code>: a chromosome end in map refinement is removed if it has a distance gap &gt; stripdis (centiMorgan) and it contains less than 5% markers.</p><p><code>maxepsilon::Real=0.5</code>: markers in map refinement are removed it they have error rates &gt; maxepsilon.</p><p><code>skeletonsize::Integer=50</code>: the number of markers in the skeleton map that is used to re-scale inter-map distances.</p><p><code>missingstring::AbstractString=&quot;NA&quot;</code>: string code for missing value.</p><p><code>isplot::Bool=false</code>: if true, plot condprob for all offspring and save in the folder &quot;outstem_plots&quot;.</p><p><code>outstem::Union{Nothing,AbstractString}=&quot;outstem&quot;</code>: stem of output filenames. If nothing, no output files.</p><p><code>logfile::Union{Nothing,AbstractString,IO}= (outstem===nothing ? nothing : string(outstem,&quot;.log&quot;))</code>: log file or IO for writing log. If nothing, no log file.</p><p><code>workdir::AbstractString = pwd()</code>: directory for reading and writing files.</p><p><code>verbose::Bool=true</code>: if true, print messages on console.</p><p><strong>Example</strong></p><pre><code class="language-julia-repl">julia&gt; polygeno = readPolyGeno(genofile,pedfile)
julia&gt; polyOrigin!(polygeno)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/chaozhi/PolyOrigin.jl/blob/906b6a5286692924a9e2b7f2c05060d964e3bebc/src/polyorigin_fun.jl#LL169-L273">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PolyOrigin.readPolyGeno" href="#PolyOrigin.readPolyGeno"><code>PolyOrigin.readPolyGeno</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">readPolyGeno(genofile, pedfile, keyargs...)</code></pre><p>reads input files and returns polygeno::PolyGeno.</p><p><strong>Positional arguments</strong></p><p><code>genofile::AbstractString</code>: filename for genotypic data. For example, a CSV formatted genofile looks like</p><pre><code class="language-none">    marker, chromosome, pos, ind1,ind2, ind3, ...
    snp1, 1, 0.14, 0, 2, 1, 4, ...
    snp2, 1, 0.16, 4, 0, NA, 2, ...
    snp3, 1, 0.21, NA, 3, 0, 1, ...</code></pre><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><ul><li><p>The first three columns specify the genetic map. Marker IDs in column 1 must be unique, chromosome IDs in column 2 must be consecutive, and positions (in unit of centi-Morgan or base pair) of markers in column 3 must be non-descreasing within a chromosome.</p></li><li><p>The rest of columns give the genotypes of sampled individuals. The indvidual IDs must be unique. The genotypes of all parents must be represented by one of the following formats 1-4, and the genotypes of all offspring must be represented by one of the following formats 1-3.</p><ol><li><code>dosage</code>: ranges from 0, 1, ..., ploidy, and NA for missing dosage;</li><li><code>readcount</code>: c1|c2, where c1 and c2 are the number of reads for alleles 1 and 2, respectively. Missing genotypesare given by 0|0</li><li><code>probability</code>: p(0)|p(1)|...|p(ploidy), where p(i) denotes the probability  of observed data given dosage i = 0, ..., ploidy, and the probabilities  are normalized so that their sum is 1.</li><li><code>phasedgeno</code>: g(1)|g(2)|...|g(ploidy),  where g(i)=1 or 2 for i=1,..., ploidy.</li></ol></li><li><p>All individuals must be in the <code>pedfile</code>.</p></li></ul></div></div><p><code>pedfile::AbstractString</code>: filename for pedigree information. For example, a CSV formatted pedfile looks like</p><pre><code class="language-none">    individual, population, motherid, fatherid, ploidy
    P1, 0, 0, 0, 4
    P3, 0, 0, 0, 4
    P3, 0, 0, 0, 4
    offspring1, pop1, P1, P2, 4
    offspring2, pop1, P1, P2, 4
    offspring3, pop2, P1, P3, 4
    offspring4, pop2, P1, P3, 4
    offspring5, pop3, P2, P3, 4
    offspring6, pop4, P3, P3, 4</code></pre><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><ul><li><p>The pedigree contains three founders (parents), two offspring from the cross beween parents 1 and 2, two offspring from the cross between parents 1 and 3, one offspring from the cross between parents 2 and 3, and one offspring from the selfing of parent 3.</p></li><li><p>All individual IDs in column 1 must be unique, column 2 denotes ID for the founder population and IDs for each F1 cross or selfing, columns 3 and 4 denotes the parents of each sub-population (motherID and fatherID of founders are set to 0), and column 5 denotes the ploidy level.</p></li><li><p>All parents and all offspring must be in the <code>genofile</code>.</p></li></ul></div></div><p><strong>Keyword arguments</strong></p><p><code>missingstring::AbstractString=&quot;NA&quot;</code>: string code for missing value.</p><p><code>delimchar::AbstractChar=&#39;,&#39;</code>:  text delimiter.</p><p><code>commentstring::AbstractString=&quot;#&quot;</code>: rows that begins with commentstring will be ignored.</p><p><code>isphysmap::Bool=false</code>: true if input markermap is physical map, where marker locations are in unit of base pair(bp).</p><p><code>recomrate::Real=1</code> recombination rate in unit of 1 cM/Mbp (centiMorgan per million base pair).</p><p><code>workdir::AbstractString = pwd()</code>: directory for reading genofile and pedfile</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/chaozhi/PolyOrigin.jl/blob/906b6a5286692924a9e2b7f2c05060d964e3bebc/src/polygeno.jl#LL94-L173">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PolyOrigin.readPolyAncestry" href="#PolyOrigin.readPolyAncestry"><code>PolyOrigin.readPolyAncestry</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">readPolyAncestry(genoprobfile, pedfile, missingstring=&quot;NA&quot;,workdir=pwd())</code></pre><p>return a struct of with type PolyAncestry from  genoprobfile and pedfile     in the directory workdir.</p><p><strong>Positional argument</strong></p><p><code>genoprobfile</code>: file storing conditional genotype probability.</p><p><code>pedfile::AbstractString</code>: filename for pedigree information.</p><p><strong>Keyword arguments</strong></p><p><code>chrpairing::Integer=44</code>: chromosome pairing that has been used in producing genoprobfile. chrpairing = 22 indicates only bivalent formations and 44 for bivalent and quadrivalent formations.</p><p><code>missingstring::AbstractString=&quot;NA&quot;</code>: string code for missing value.</p><p><code>delimchar::AbstractChar=&#39;,&#39;</code>:  text delimiter.</p><p><code>commentstring::AbstractString=&quot;#&quot;</code>: rows that begins with commentstring will be ignored.</p><p><code>workdir::AbstractString = pwd()</code>: directory for reading genoprobfile and pedfile</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/chaozhi/PolyOrigin.jl/blob/906b6a5286692924a9e2b7f2c05060d964e3bebc/src/polyancestry.jl#LL299-L325">source</a></section><section><div><pre><code class="language-none">readPolyAncestry(ancestryfile, missingstring=&quot;NA&quot;,workdir=pwd())</code></pre><p>return a struct of type PolyAncestry from  ancestryfile in the     directory workdir.</p><p><strong>Positional argument</strong></p><p><code>ancestryfile</code>: file storing polyancestry that is generated by <a href="#PolyOrigin.savePolyAncestry"><code>savePolyAncestry</code></a> It is one of output files by <a href="#PolyOrigin.polyOrigin"><code>polyOrigin</code></a>.</p><p><strong>Keyward arguments</strong></p><p><code>missingstring::AbstractString=&quot;NA&quot;</code>: string code for missing value.</p><p><code>workdir::AbstractString = pwd()</code>: directory for reading ancestryfile,</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/chaozhi/PolyOrigin.jl/blob/906b6a5286692924a9e2b7f2c05060d964e3bebc/src/polyancestry.jl#LL393-L410">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PolyOrigin.savePolyAncestry" href="#PolyOrigin.savePolyAncestry"><code>PolyOrigin.savePolyAncestry</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">savePolyAncestry(outfile,polyancestry,missingstring=&quot;NA&quot;,workdir=pwd())</code></pre><p>saves polyancestry into outfile in CSV format.</p><p><strong>Positional arguments</strong></p><p><code>outfile::AbstractString</code>: file for saving polyancestry in the directory workdir. The saved outfile consists of several tables: the first row of each table has two cells: [PolyOrigin-PolyAncestry, nameoftable], and the rest rows denote a dataframe with column names. The following is a list of table names and their descriptions.</p><ol><li><p><code>designinfo</code>: design information with columns being [population, parent1, parent2, ...]. Matrix element D<em>ij denotes the nubmer of gametes contributed to each offspring in i-th population by j-th parent. D</em>ij=2 means that i-th population is produced by self-fertilization of j-th parent.</p></li><li><p><code>parentinfo</code>: parent information with columns being [individual, ploidy].</p></li><li><p><code>offspringinfo</code>: offspring information with columns being [individual, population, ploidy, isoutlier].</p></li><li><p><code>delmarker</code>: markers that were removed from markermap, in the stages of parental phasing</p></li></ol><p>or map refinement, with columns being [:marker, :chromosome, :position]</p><ol><li><code>correction</code>: parental genotype corrections with columns being</li></ol><p>[:round, :marker,:chromosome,:parent,:old<em>genotype,:new</em>genotype,:old<em>nerr,:new</em>nerr].</p><ol><li><p><code>valentlist</code>: list of possible bi- or multi-valent formations for each population. The dataframe has columns [population, parentindex, parent, valentindex, valent].</p></li><li><p><code>valentprob</code>: posterior valent probability for each offspring in each chromosome. The dataframe has columns [chromosome, individual, population, valentindex, valent, loglike, valentprob].</p></li><li><p><code>parentgeno</code>: phased parental genotypes. The dataframe has columns [marker, chromosome, position, parent1, parent2, ...].</p></li><li><p><code>ancestralgenotype</code>: list of origin-genotypes for each population. The dataframe has columns [population, parentindex, parent, stateindex, state]</p></li><li><p><code>genoprob</code>: marginal posterior probabilities for origin-genotypes. For each offspring at each marker, the posterior probability vector is represented by a sparse vector in the form of I=&gt;V: state index vector I and non-zero probability vector V such that I[k]=V[k] for k=1, ..., K. The elements of vector I or V are delimited by &quot;|&quot;.</p></li></ol><p><code>polyancestry::PolyAncestry</code>: results of haplotype reconstruction returned  from <a href="#PolyOrigin.polyOrigin"><code>polyOrigin</code></a>.</p><p><strong>Keyward arguments</strong></p><p><code>missingstring::AbstractString=&quot;NA&quot;</code>: string code for missing value.</p><p><code>workdir::AbstractString = pwd()</code>: directory for writing outfile.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/chaozhi/PolyOrigin.jl/blob/906b6a5286692924a9e2b7f2c05060d964e3bebc/src/polyancestry.jl#LL103-L157">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PolyOrigin.plotMapComp" href="#PolyOrigin.plotMapComp"><code>PolyOrigin.plotMapComp</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">plotMapComp(mapx, mapy,
    boundaryline = (1.5,:dot,:black),
    plotmarker=(:star, 5, 0.5,:blue,stroke(:blue)),
    isannotate= true,
    xlabel=&quot;mapx position&quot;,
    ylabel=&quot;mapy position&quot;)</code></pre><p>plot postions of mapx vs those of mapx.</p><p><strong>Positional arguments</strong></p><p><code>mapx::Vector{DataFrame}</code>: marker map for all chromosomes.</p><p><code>mapy::Vector{DataFrame}</code>: comparing map for all chromosomes.</p><p><strong>Keyword arguments</strong></p><p><code>boundaryline=(1.5,:dot,:black)</code>: vertical lines for chromosome boundaries.</p><p><code>plotmarker=(:star, 5, 0.5,:blue,stroke(:blue))</code>: scatter markers.</p><p><code>isannotate::Bool=true</code>: if ture,  annotate chromosome ID and kendall correlation.</p><p><code>xlabel::AbstractString=&quot;mapx position&quot;</code>: axis x label</p><p><code>ylabel::AbstractString=&quot;mapy position&quot;</code>: axis y label</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/chaozhi/PolyOrigin.jl/blob/906b6a5286692924a9e2b7f2c05060d964e3bebc/src/polyplot.jl#LL247-L275">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PolyOrigin.plotCondprob" href="#PolyOrigin.plotCondprob"><code>PolyOrigin.plotCondprob</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">plotCondprob(polyancestry,offspring=nothing,ishaploprob=true,
    colorgradient = ColorGradient([:white,:blue,:red]),
    boundaryline = (1.5,:dot,:black),
    truemarker=(:star, 5, 0.5,:gray,stroke(:gray)),
    truegeno=nothing)</code></pre><p>plot heatmap for conditional probability.</p><p><strong>Positional arguments</strong></p><p><code>polyancestry::PolyAncestry</code>: polyancestry returned from <a href="#PolyOrigin.polyOrigin"><code>polyOrigin</code></a>.</p><p><strong>Keyword arguments</strong></p><p><code>offspring::Union{Nothing,Integer}=nothing</code>: offsprign index and random offspring by default.</p><p><code>colorgradient::ColorGradient=ColorGradient([:white,:blue,:red])</code>: color gradient for heatmap</p><p><code>left_margin = :match</code>: Specifies the extra padding to the left of the subplot</p><p><code>bottom_margin = :match</code>: Specifies the extra padding to the bottom of the subplot</p><p><code>boundaryline=(1.5,:dot,:black)</code>: vertical lines for chromosome boundaries.</p><p><code>truemarker=(:star, 5, 0.5,:gray,stroke(:gray))</code>: scatter markers for true ancestral states.</p><p><code>truegeno::Union{Nothing,NamedTuple}</code>: provides true parental origins. It is generated by <a href="#PolyOrigin.readTruegeno!"><code>readTruegeno!</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/chaozhi/PolyOrigin.jl/blob/906b6a5286692924a9e2b7f2c05060d964e3bebc/src/polyplot.jl#LL1-L34">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PolyOrigin.animCondprob" href="#PolyOrigin.animCondprob"><code>PolyOrigin.animCondprob</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">animCondprob(polyancestry,fps=1,outfile=nothing,kewargs...)</code></pre><p>animation for plots of conditional probability.</p><p><strong>Positional arguments</strong></p><p><code>polyancestry::PolyAncestry</code>: polyancestry returned from <a href="#PolyOrigin.polyOrigin"><code>polyOrigin</code></a>.</p><p><strong>Keyword arguments</strong></p><p><code>fps::Real=1</code>: number of frames per seconds.</p><p><code>outfile::AbstractString=&quot;condprob.gif&quot;</code>: output file for saving animation.</p><p>see <a href="#PolyOrigin.plotCondprob"><code>plotCondprob</code></a> for keyargs.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/chaozhi/PolyOrigin.jl/blob/906b6a5286692924a9e2b7f2c05060d964e3bebc/src/polyplot.jl#LL130-L147">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PolyOrigin.readTruegeno!" href="#PolyOrigin.readTruegeno!"><code>PolyOrigin.readTruegeno!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">readTruegeno!(truefile,polyancestry, keyargs...)</code></pre><p>returned truegeno of NamedTuple with truegeno.parentgeno storing phased parental genotypes, and truegeno.offspringgeno storing true parental origins. <code>polyancestry</code> is modified (1) extra markers in truefile are put into polyancestry.delmarker, (2) absolute phase of polyancestry.parentgeno is set according to truegeno.parentgeno, and (3) polyancestry.genoprob and polyancestry.haploprob are set to be consistent with the absolute phase.</p><p><strong>Positional arguments</strong></p><p><code>truefile::AbstractString</code>: file storing true values, which has the same requirement as the <code>genofile</code>. The parental genotypes must be given in the <code>phasedgeno</code>  format with alleles being &quot;1&quot; and &quot;2&quot;. The offspring origin-genotypes must be given in the <code>phasedgeno</code>  format with allels being parental homologs.</p><p><code>polyancestry::PolyAncestry</code>: for checking the consistency of marker IDs, chromosome IDs, and individual IDs.</p><p><strong>Keyword arguments</strong></p><p><code>delimchar::AbstractChar=&#39;,&#39;</code>:  text delimiter.</p><p><code>commentstring::AbstractString=&quot;#&quot;</code>: rows that begins with commentstring will be ignored.</p><p><code>workdir::AbstractString = pwd()</code>: directory for reading truevalue file.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/chaozhi/PolyOrigin.jl/blob/906b6a5286692924a9e2b7f2c05060d964e3bebc/src/calaccuracy.jl#LL3-L31">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="PolyOrigin.calAccuracy!" href="#PolyOrigin.calAccuracy!"><code>PolyOrigin.calAccuracy!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">calAccuracy!(truefile,polyancestry,workdir=pwd(),io=nothing,verbose=true)</code></pre><p>calculate phasing and ancestral inference accuracies accoring to the true values in <code>truefile</code>. The <code>polyancestry</code> is modified with absolute parental phases and consistent genoprob.</p><p><strong>Positional arguments</strong></p><p><code>truefile::AbstractString</code>: contains true values for parental phases and ancestral genotypes.</p><p><code>polyancestry::PolyAncestry</code>: a struct returned by <a href="#PolyOrigin.polyOrigin"><code>polyOrigin</code></a>.</p><p><strong>Keyward arguments</strong></p><p><code>workdir::AbstractString = pwd()</code>: directory for writing outfile.</p><p><code>io::Union{Nothing,IOStream}</code>: stream for writing log.</p><p><code>verbose::Bool=true</code>: true if print messages on console.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/chaozhi/PolyOrigin.jl/blob/906b6a5286692924a9e2b7f2c05060d964e3bebc/src/calaccuracy.jl#LL160-L182">source</a></section><section><div><pre><code class="language-none">calAccuracy!(truegeno,polyancestry,io=nothing,verbose=true)</code></pre><p>calculate phasing and ancestral inference accuracies accoring to the true values in <code>truegeno</code>. The <code>polyancestry</code> is modified with absolute parental phases and consistent genoprob.</p><p><strong>Positional arguments</strong></p><p><code>truegeno::NamedTuple</code>: contains true values for parental phases and ancestral genotypes, returned by <a href="#PolyOrigin.readTruegeno!"><code>readTruegeno!</code></a>.</p><p><code>polyancestry::PolyAncestry</code>: a struct returned by <a href="#PolyOrigin.polyOrigin"><code>polyOrigin</code></a>.</p><p><strong>Keyward arguments</strong></p><p><code>io::Union{Nothing,IOStream}</code>: stream for writing log.</p><p><code>verbose::Bool=true</code>: true if print messages on console.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/chaozhi/PolyOrigin.jl/blob/906b6a5286692924a9e2b7f2c05060d964e3bebc/src/calaccuracy.jl#LL190-L210">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../../man/3_examples/">« 3 Examples</a><a class="docs-footer-nextpage" href="../internals/">Internals »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Thursday 29 April 2021 07:17">Thursday 29 April 2021</span>. Using Julia version 1.5.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
